File: src/java/org/orbeon/oxf/xforms/control/controls/XFormsRangeControl.java
/**
 * Copyright (C) 2010 Orbeon, Inc.
 *
 * This program is free software; you can redistribute it and/or modify it under the terms of the
 * GNU Lesser General Public License as published by the Free Software Foundation; either version
 * 2.1 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * The full text of the license is available at http://www.gnu.org/copyleft/lesser.html
 */
package org.orbeon.oxf.xforms.control.controls;

import org.dom4j.Element;
import org.orbeon.oxf.xforms.XFormsConstants;
import org.orbeon.oxf.xforms.control.XFormsControl;
import org.orbeon.oxf.xforms.control.XFormsValueControl;
import org.orbeon.oxf.xforms.xbl.XBLContainer;
import org.orbeon.oxf.xml.XMLConstants;

import java.util.Map;

/**
 * Represents an xforms:range control.
 */
public class XFormsRangeControl extends XFormsValueControl {

    private String start;
    private String end;
    private String step;

    public XFormsRangeControl(XBLContainer container, XFormsControl parent, Element element, String name, String id, Map<String, Element> state) {
        super(container, parent, element, name, id);
        this.start = element.attributeValue("start");
        this.end = element.attributeValue("end");
        this.step = element.attributeValue("step");
    }

    @Override
    public boolean hasJavaScriptInitialization() {
        return true;
    }

    public String getStart() {
        return start;
    }

    public String getEnd() {
        return end;
    }

    public String getStep() {
        return step;
    }

    @Override
    public void storeExternalValue(String value, String type) {
        // Store after converting
        super.storeExternalValue(convertFromExternalValue(value), type);
    }

    private String convertFromExternalValue(String externalValue) {
        final String typeName = getBuiltinTypeName();
        if (getStart() != null && getEnd() != null && "integer".equals(typeName)) {
            final int start = Integer.parseInt(getStart());
            final int end = Integer.parseInt(getEnd());

            final int value = start + ((int) (Double.parseDouble(externalValue) * (double) (end - start)));
            return Integer.toString(value);
        } else {
            return externalValue;
        }
    }

    @Override
    protected void evaluateExternalValue() {
        final String internalValue = getValue();
        final String updatedValue;
        if (internalValue == null) {// can it be really?
            updatedValue = null;
        } else if (getStart() != null && getEnd() != null
                && (XMLConstants.XS_INTEGER_QNAME.equals(getType()) || XFormsConstants.XFORMS_INTEGER_QNAME.equals(getType()))) {

            final int start = Integer.parseInt(getStart());
            final int end = Integer.parseInt(getEnd());

            final double value = ((double) (Integer.parseInt(internalValue) - start)) / ((double) end - start);
            updatedValue = Double.toString(value);
        } else {
            updatedValue = internalValue;
        }
        setExternalValue(updatedValue);
    }
}

File: xforms/jvm/src/main/scala/org/orbeon/oxf/xforms/control/controls/XFormsRangeControl.scala
/**
 * Copyright (C) 2016 Orbeon, Inc.
 *
 * This program is free software; you can redistribute it and/or modify it under the terms of the
 * GNU Lesser General Public License as published by the Free Software Foundation; either version
 * 2.1 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * The full text of the license is available at http://www.gnu.org/copyleft/lesser.html
 */
package org.orbeon.oxf.xforms.control.controls

import org.orbeon.dom.Element
import org.orbeon.xforms.XFormsConstants._
import org.orbeon.oxf.xforms.control.{XFormsControl, XFormsSingleNodeControl, XFormsValueControl}
import org.orbeon.oxf.xforms.xbl.XBLContainer
import org.orbeon.oxf.xml.XMLConstants._
import org.orbeon.saxon.om.Item

// Represent an xf:range control.
// NOTE: This should instead be implemented with an XBL component.
class XFormsRangeControl(
  container   : XBLContainer,
  parent      : XFormsControl,
  element     : Element,
  effectiveId : String
) extends XFormsSingleNodeControl(
  container,
  parent,
  element,
  effectiveId
) with XFormsValueControl {

  private val start = element.attributeValue("start")
  private val end   = element.attributeValue("end")
//  private val step  = element.attributeValue("step")

  override def hasJavaScriptInitialization = true

  override def translateExternalValue(boundItem: Item, externalValue: String): Option[String] =
    Option(convertFromExternalValue(externalValue))

  private def convertFromExternalValue(externalValue: String) = {
    if (start != null && end != null && getBuiltinTypeName == "integer") {
      val startInt = start.toInt
      val endInt   = end.toInt
      val value    = startInt + (externalValue.toDouble * (endInt - startInt).toDouble).toInt

      value.toString
    } else {
      externalValue
    }
  }

  override def evaluateExternalValue(): Unit = {
    val internalValue = getValue

    val updatedValue =
      if (internalValue == null) {
        null
      } else if (start != null && end != null && (valueType == XS_INTEGER_QNAME || valueType == XFORMS_INTEGER_QNAME)) {
        val startInt = start.toInt
        val endInt   = end.toInt
        val value     = (internalValue.toInt - startInt).toDouble / (endInt.toDouble - startInt)

        value.toString
      } else {
        internalValue
      }
    setExternalValue(updatedValue)
  }

  override def findAriaByControlEffectiveId = None
}